addons/udes_stock_routing/models/stock_rule.py

```diff
@@ -0,0 +1,44 @@
+from odoo import api, fields, models
+
+
+class StockRule(models.Model):
+    _inherit = "stock.rule"
+
+    u_push_on_drop = fields.Boolean(
+        "Push on Drop-off",
+        default=False,
+        help="When stock is dropped in the "
+        "from_location of this path, "
+        "automatically push it onwards.",
+    )
+
+    @api.model
+    def get_path_from_location(self, location):
+        """
+        Find a single stock.rule for which the given location is
+        a valid starting location. If one not found return an empty recordset.
+        """
+        push_steps = self.env["stock.rule"].search(
+            [
+                (
+                    "location_src_id",
+                    "parent_of",
+                    [
+                        location.id,
+                    ],
+                ),
+                ("u_push_on_drop", "=", True),
+            ]
+        )
+        if push_steps:
+            return push_steps.sorted(key=lambda p: p.location_src_id.parent_path.strip("/").split("/")[-1], reverse=True)[0]
```

Comment:
Can we add some doc strings to this please. As i can't tell what is going on, parent_path is a char field, and we format it to the last part of the path, assuming the path follows that format ( don't know if there is a constraint on it or not). Then we choose the last record sorted by the path name? 
This is just a question, i haven't used this method before so i don't understand why we choose that particular record if there are more.

Reply:
Sorry I am still trying to work it all out. This is how I thought it worked before commiting this code: If there are two paths defined at different levels of the location hierarchy, for example: if we have a location hierarchy of Check, Check/CheckzoneA and Check/CheckzoneA/droplocation. When stock is dropped in Check/CheckzoneA/droplocation this method searches for the stock.rules (paths) attached to Check and CheckzoneA, if there are two paths defined on this location for example:
Check has parent_path "1/24/"
CheckzoneA has parent_path "1/24/25/"
(From what I can tell the string is always formatted in this way)
Then we sort by taking only the last number in the parent path. So after stripping and splitting:
Check would be represented by 24
CheckzoneA would be represented by 25 
We want the largest number as it is the more specific path. That why the numbers are sorted by descending order as opposed to ascending order. 

However after checking the test cases again (the test case test_many_routes), I think that in Odoo 14 it will only return the closest path e.g. it will only return CheckzoneA with parent_path "1/24/25/" and not Check with parent_path "1/24/". This is different to what happened in Odoo 11 where both parent locations where returned (in fact Odoo 11 used a completely different variable to parent path called parent_left, see: https://www.odoo.com/forum/help-1/what-is-the-meaning-of-parent-left-and-parent-right-domain-operator-25221). The sorting might not be needed when dealing with different stock.rules (paths) at different locations on the hierarchy. But, if there are multiple stock.rules defined on the same location, e.g. if we have two stock.rules defined on CheckzoneA, then I am still thinking about which stock location to choose? As it stands the newest stock.rule will be chosen. 

Perhaps there should be a flag to indicate which stock.rule should be used when two stock.rules are specified on the same location?   

I hope that all makes sense.

Reply:
Can we just update the doc string - i understand now :)